File: code/data_structures/fenwick_tree/FenwickTreeRangeQueryPointUpdate.py


class FenwickTreeRangeQueryPointUpdate:
    def __init__(self, sz_or_values):
        if isinstance(sz_or_values, int):
            self.N = sz_or_values + 1
            self.tree = [0] * self.N
        elif isinstance(sz_or_values, list):
            if sz_or_values is None:
                raise ValueError("Values array cannot be None!")
            
            self.N = len(sz_or_values)
            sz_or_values[0] = 0
            
            self.tree = sz_or_values[:]
            
            for i in range(1, self.N):
                parent = i + self.lsb(i)
                if parent < self.N:
                    self.tree[parent] += self.tree[i]
        else:
            raise ValueError("Invalid argument type for constructor")
    
    def lsb(self, i):
        return i & -i
    
    def prefix_sum(self, i):
        sum_val = 0
        while i != 0:
            sum_val += self.tree[i]
            i &= ~self.lsb(i)
        return sum_val
    
    def sum(self, left, right):
        if right < left:
            raise ValueError("Make sure right >= left")
        return self.prefix_sum(right) - self.prefix_sum(left - 1)
    
    def get(self, i):
        return self.sum(i, i)
    
    def add(self, i, v):
        while i < self.N:
            self.tree[i] += v
            i += self.lsb(i)
    
    def set(self, i, v):
        self.add(i, v - self.sum(i, i))
    
    def __str__(self):
        return str(self.tree)

File: code/data_structures/fenwick_tree/FenwickTreeRangeQueryPointUpdate.cpp


#include <vector>
#include <iostream>

class FenwickTreeRangeQueryPointUpdate {
private:
    int N;
    std::vector<long long> tree;

public:
    FenwickTreeRangeQueryPointUpdate(int sz) : N(sz + 1), tree(N, 0) {}

    FenwickTreeRangeQueryPointUpdate(std::vector<long long>& values) {
        N = values.size();
        values.insert(values.begin(), 0);
        tree = values;

        for (int i = 1; i < N; ++i) {
            int parent = i + lsb(i);
            if (parent < N) tree[parent] += tree[i];
        }
    }

    int lsb(int i) {
        return i & -i;
    }

    long long prefixSum(int i) {
        long long sum = 0LL;
        while (i != 0) {
            sum += tree[i];
            i &= ~lsb(i);
        }
        return sum;
    }

    long long sum(int left, int right) {
        if (right < left) throw std::invalid_argument("Make sure right >= left");
        return prefixSum(right) - prefixSum(left - 1);
    }

    long long get(int i) {
        return sum(i, i);
    }

    void add(int i, long long v) {
        while (i < N) {
            tree[i] += v;
            i += lsb(i);
        }
    }

    void set(int i, long long v) {
        add(i, v - sum(i, i));
    }

    void print() {
        for (int i = 1; i < N; ++i) {
            std::cout << tree[i] << " ";
        }
        std::cout << std::endl;
    }
};
